### ISSUE TITLE
mcrypt() key length is enforced starting with PHP 5.6

### ISSUE THREAD

**Issue by vaindil** [bemosior/PHPasswordPusher#31]
As of PHP 5.6.0, the [`mcrypt_encrypt()`](https://php.net/manual/en/function.mcrypt-encrypt.php) and [`mcrypt_decrypt()`](https://php.net/manual/en/function.mcrypt-decrypt.php) functions no longer accept key values that are inconsistent with the chosen cipher (`MCRYPT_RIJNDAEL_256` in this case). No visible error is ever displayed by the program--everything works correctly, but when accessing the generated link no value is displayed because it a) was never encrypted/stored in the database and therefore b) can't actually be retrieved/decrypted.

In order for `mcrypt` to work properly with this cipher, the supplied key has to be 16, 24, or 32 characters long. Not sure how you'd like to handle this, but that's the problem in 5.6.0 on.

I changed my key and salt when the problem started to avoid security issues (which is why they're short), but these are the relevant error messages from my apache log:

    [:error] [pid 2017] [client <hidden>] PHP Warning:  mcrypt_encrypt(): Key of size 11 not supported by this algorithm. Only keys of sizes 16, 24 or 32 supported in /var/www/phpwp/pwpusher_private/security.php on line 28, referer: https://example/pwpusher/
    [:error] [pid 2022] [client <hidden>] PHP Warning:  mcrypt_decrypt(): Key of size 11 not supported by this algorithm. Only keys of sizes 16, 24 or 32 supported in /var/www/phpwp/pwpusher_private/security.php on line 55

**Comment by bemosior** [comment]
First, I'd like to re-evaluate MCRYPT_RIJNDAEL_256 for this purpose (it's probably still a reasonable choice, but I would like to do some reading on the subject). 

Second, I think that good application behavior here would be to truncate the key to the next-lowest appropriate length so that the system still functions. Inexplicably non-functional systems are just frustrating to work with, though I also entertain the argument that a system should fail-closed instead of performing an operation to make it less secure (i.e., truncating the key).

If there's adequate documentation in the configuration file and an error hits the log every time that operation occurs, I think that's a sufficiently responsible outcome. This should allow a person to make an educated decision to choose a specific key size

**Comment by vaindil** [comment]
All good points. This is why I made an issue for it rather than a pull request like with my last edits.

I found [this post](https://defuse.ca/secure-php-encryption.htm) that has some good information and links to [this repo](https://github.com/defuse/php-encryption) symmetric encryption library. It looks pretty good to me, though it would bump PHPWP's required PHP version from 5.3 to 5.4. (Not a big deal to me, seeing how 5.3 isn't officially supported anymore, but that's your call.)

I think truncating the length could work, but in my test example I just used a couple words that only came out to 11 characters. Truncating wouldn't work in that case and I'm not a fan of arbitrarily padding it somehow if it's too short (introduces too many potential issues in my opinion; a solution would need to be used such as storing a DB variable that's always queried and appended to the key).

I think the best option would be to add a check for the `mcrypt` return value (either a proper output or `FALSE`), and just die immediately if it fails, printing a simple "Invalid key" error message. I know not everyone will use a key length that's exactly correct based on the algorithm even if it's in the documentation, but I would _hope_ that anyone using this would at least check that their installation works. The error message doesn't give any security information away other than the fact that the key (in this case) isn't 16, 24, or 32 characters long (which isn't much information at all), and it would prompt them to check the documentation or the readme again to see why that error message is appearing. The error could even be changed to "Check your log file", which gives away _no_ security information.

Basically, I don't see the harm in having this fail loudly--the bug would already be visible to any users who encounter it because the returned credential would be blank, so simply displaying nothing on the screen other than "Invalid key" or "Check your log file" shouldn't be an issue.

**Comment by bemosior** [comment]
Thanks for the links. I'll do some reading as soon as I can, though I should mention that I'm hesitant to dump support for php 5.3, since 5.4 isn't available by default in CentOS 6 (and I imagine RHEL 6). 

You've provided a very good argument for loud failure. Let's go that route. I'll try to catch up on the options later this weekend or next week and report back here before making any major changes.

I've added you as a collaborator to the project. All I ask is that we keep new work contained on the develop (or in another feature) branch for now. We can PR from `develop` or other feature branches --> `master` when we're satisfied with the changes.

**Comment by vaindil** [comment]
Well, I wasn't expecting that. Yes, I'll only modify `develop` and we can go from there. One question about environment requirements--the project currently lists support for CentOS 5, which I'll admit I've never worked with before (I started with 6). I have no problem with that, since 5 is still being maintained through 2017, but I'm going to reimage a VPS for testing (I don't use it anyway, figure I may as well have one on Ubuntu and one on CentOS). Should I go back to CentOS 5 to test against that or are we going to move up to 6?

**Comment by bemosior** [comment]
Yeah let's keep supporting CentOS 5 if possible until it's dropped, but we should probably test 6 as well. I can certainly take care of that part, if you'd like. I really like the local environment paradigm provided by Vagrant. Looks like there are a couple existing boxes for CentOS already: https://atlas.hashicorp.com/boxes/search?utf8=%E2%9C%93&sort=&provider=&q=centos

**Comment by vaindil** [comment]
No problem. After I posted that I discovered my provider actually only has 6.4 and up available to image, so I imaged 6.6 and will install a local copy of 5 for testing as well.

**Comment by bemosior** [comment]
Based on these articles and others, I learned that with `MCRYPT_RIJNDAEL_256`, the `256` refers to the block size, not the key length. `MCRYPT_RIJNDAEL_128` is essentially `AES`, but both `MCRYPT_RIJNDAEL_128` and `MCRYPT_RIJNDAEL_256` can use `128`, `192`, or `256`-bit keys. `AES-128/192/256` does actually refer to the key length.

By using `MCRYPT_RIJNDAEL_256`, we're not doing anything wrong per se (just using a different block length), but it might be ok to switch to `MCRYPT_RIJNDAEL_128` just so we can say we use `AES`. Thoughts?


After reading [this article](http://www.leaseweblabs.com/2014/02/aes-php-mcrypt-key-padding/), I learned  that mcrypt previously would pad the key up to the next `128`, `192`, or `256`-bit length.

Now I'm coming full-circle with your original bug report, which points out that with PHP 5.6.0, mcrypt has [changed the behavior](http://php.net/manual/en/function.mcrypt-encrypt.php) to require exact-length keys:
>Invalid key and iv sizes are no longer accepted. mcrypt_encrypt() will now throw a warning and return FALSE if the inputs are invalid. Previously keys and IVs were padded with '\0' bytes to the next valid size.

Now that I (more or less) comprehend the full context, I believe we should maintain mcrypt's behavior and fail hard when the key length isn't exact. The person setting up the system should make a deliberate choice between `AES-128`, `AES-192`, or `AES-256` by using a key of the appropriate length. What do you think?

**Comment by bemosior** [comment]
It also looks like the project uses `MCRYPT_RAND` for IV creation, and the [current default](http://php.net/manual/en/function.mcrypt-create-iv.php) is `MCRYPT_DEV_URANDOM`. I think we could make that switch, since the randomness should be better than `MCRYPT_RAND` and nonblocking, as opposed to `MCRYPT_DEV_RANDOM`.

**Comment by vaindil** [comment]
Sorry for the the very late reply, I should be able to reply much more quickly from now on!

I agree with using `MCRYPT_RIJNDAEL_128`. It doesn't take any work to make the change, and saying we use `AES` is more publicly recognizable than otherwise for anyone who cared to take our word for it (though anyone who _really_ cares will look into the source code anyway, but still). I don't think we're going to be hitting a problem with the cap of 256 exabytes, so we should be good there.

I also agree with failing hard to require exact-length keys--including the requirements in the documentation isn't difficult, and the short, cryptic message on failure isn't a security risk, so I believe that would be the best route to go. I also agree with using `MCRYPT_DEV_URANDOM`; from my own research `/dev/urandom` is a better source of randomness than `/dev/random` (plus it's nonblocking, as you mentioned).

TL;DR: I agree.

**Comment by bemosior** [comment]
Don't apologize! I'll see what I can do to get started.

**Comment by bemosior** [comment]
I found another issue with the implementation regarding the IV and the mcrypt mode: #37

**Comment by bemosior** [comment]
@vaindil With #37 created as a separate issue, I think this particular issue can be closed. Did I miss anything?

**Comment by vaindil** [comment]
Looks good to me.

**Comment by bemosior** [comment]
Thanks, @vaindil, for everything you've been doing! :+1:

**Comment by Bobcui001** [comment]
What are the effects of the version?

**Comment by bemosior** [comment]
@Bobcui001, could you clarify the question? The version of what?
